{
  "cells": [
    {
      "cell_type": "code",
      "execution_count": 1,
      "id": "3d7f1326",
      "metadata": {
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "Loading required package: SeuratObject\n",
            "\n",
            "Loading required package: sp\n",
            "\n",
            "\n",
            "Attaching package: ‘SeuratObject’\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:base’:\n",
            "\n",
            "    intersect, t\n",
            "\n",
            "\n",
            "\n",
            "Attaching package: ‘dplyr’\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:stats’:\n",
            "\n",
            "    filter, lag\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:base’:\n",
            "\n",
            "    intersect, setdiff, setequal, union\n",
            "\n",
            "\n",
            "── \u001b[1mAttaching core tidyverse packages\u001b[22m ──────────────────────── tidyverse 2.0.0 ──\n",
            "\u001b[32m✔\u001b[39m \u001b[34mforcats  \u001b[39m 1.0.0     \u001b[32m✔\u001b[39m \u001b[34mstringr  \u001b[39m 1.6.0\n",
            "\u001b[32m✔\u001b[39m \u001b[34mlubridate\u001b[39m 1.9.4     \u001b[32m✔\u001b[39m \u001b[34mtibble   \u001b[39m 3.3.0\n",
            "\u001b[32m✔\u001b[39m \u001b[34mpurrr    \u001b[39m 1.1.0     \u001b[32m✔\u001b[39m \u001b[34mtidyr    \u001b[39m 1.3.1\n",
            "\u001b[32m✔\u001b[39m \u001b[34mreadr    \u001b[39m 2.1.5     \n",
            "── \u001b[1mConflicts\u001b[22m ────────────────────────────────────────── tidyverse_conflicts() ──\n",
            "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m masks \u001b[34mstats\u001b[39m::filter()\n",
            "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m    masks \u001b[34mstats\u001b[39m::lag()\n",
            "\u001b[36mℹ\u001b[39m Use the conflicted package (\u001b[3m\u001b[34m<http://conflicted.r-lib.org/>\u001b[39m\u001b[23m) to force all conflicts to become errors\n",
            "Loading required package: SingleCellExperiment\n",
            "\n",
            "Loading required package: SummarizedExperiment\n",
            "\n",
            "Loading required package: MatrixGenerics\n",
            "\n",
            "Loading required package: matrixStats\n",
            "\n",
            "\n",
            "Attaching package: ‘matrixStats’\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:dplyr’:\n",
            "\n",
            "    count\n",
            "\n",
            "\n",
            "\n",
            "Attaching package: ‘MatrixGenerics’\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:matrixStats’:\n",
            "\n",
            "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
            "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
            "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
            "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
            "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
            "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
            "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
            "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
            "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
            "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
            "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
            "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
            "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
            "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
            "    rowWeightedSds, rowWeightedVars\n",
            "\n",
            "\n",
            "Loading required package: GenomicRanges\n",
            "\n",
            "Loading required package: stats4\n",
            "\n",
            "Loading required package: BiocGenerics\n",
            "\n",
            "\n",
            "Attaching package: ‘BiocGenerics’\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:lubridate’:\n",
            "\n",
            "    intersect, setdiff, union\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:dplyr’:\n",
            "\n",
            "    combine, intersect, setdiff, union\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:SeuratObject’:\n",
            "\n",
            "    intersect\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:stats’:\n",
            "\n",
            "    IQR, mad, sd, var, xtabs\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:base’:\n",
            "\n",
            "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
            "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
            "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
            "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
            "    Position, rank, rbind, Reduce, rownames, sapply, saveRDS, setdiff,\n",
            "    table, tapply, union, unique, unsplit, which.max, which.min\n",
            "\n",
            "\n",
            "Loading required package: S4Vectors\n",
            "\n",
            "\n",
            "Attaching package: ‘S4Vectors’\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:lubridate’:\n",
            "\n",
            "    second, second<-\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:tidyr’:\n",
            "\n",
            "    expand\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:dplyr’:\n",
            "\n",
            "    first, rename\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:utils’:\n",
            "\n",
            "    findMatches\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:base’:\n",
            "\n",
            "    expand.grid, I, unname\n",
            "\n",
            "\n",
            "Loading required package: IRanges\n",
            "\n",
            "\n",
            "Attaching package: ‘IRanges’\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:lubridate’:\n",
            "\n",
            "    %within%\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:purrr’:\n",
            "\n",
            "    reduce\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:dplyr’:\n",
            "\n",
            "    collapse, desc, slice\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:sp’:\n",
            "\n",
            "    %over%\n",
            "\n",
            "\n",
            "Loading required package: GenomeInfoDb\n",
            "\n",
            "Loading required package: Biobase\n",
            "\n",
            "Welcome to Bioconductor\n",
            "\n",
            "    Vignettes contain introductory material; view with\n",
            "    'browseVignettes()'. To cite Bioconductor, see\n",
            "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
            "\n",
            "\n",
            "\n",
            "Attaching package: ‘Biobase’\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:MatrixGenerics’:\n",
            "\n",
            "    rowMedians\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:matrixStats’:\n",
            "\n",
            "    anyMissing, rowMedians\n",
            "\n",
            "\n",
            "\n",
            "Attaching package: ‘SummarizedExperiment’\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:Seurat’:\n",
            "\n",
            "    Assays\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:SeuratObject’:\n",
            "\n",
            "    Assays\n",
            "\n",
            "\n",
            "Loading required package: Matrix\n",
            "\n",
            "\n",
            "Attaching package: ‘Matrix’\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:S4Vectors’:\n",
            "\n",
            "    expand\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:tidyr’:\n",
            "\n",
            "    expand, pack, unpack\n",
            "\n",
            "\n",
            "Registered S3 method overwritten by 'pROC':\n",
            "  method   from            \n",
            "  plot.roc spatstat.explore\n",
            "\n",
            "\n",
            "Attaching package: ‘celda’\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:S4Vectors’:\n",
            "\n",
            "    params\n",
            "\n",
            "\n",
            "Loading required package: reticulate\n",
            "\n",
            "Welcome to enrichR\n",
            "Checking connections ... \n",
            "\n",
            "Enrichr ... \n",
            "Connection is Live!\n",
            "\n",
            "FlyEnrichr ... \n",
            "Connection is Live!\n",
            "\n",
            "WormEnrichr ... \n",
            "Connection is Live!\n",
            "\n",
            "YeastEnrichr ... \n",
            "Connection is Live!\n",
            "\n",
            "FishEnrichr ... \n",
            "Connection is Live!\n",
            "\n",
            "OxEnrichr ... \n",
            "Connection is Live!\n",
            "\n",
            "Loading required package: patchwork\n",
            "\n",
            "\n",
            "Attaching package: ‘magrittr’\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:GenomicRanges’:\n",
            "\n",
            "    subtract\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:purrr’:\n",
            "\n",
            "    set_names\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:tidyr’:\n",
            "\n",
            "    extract\n",
            "\n",
            "\n",
            "\n",
            "Attaching package: ‘glue’\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:SummarizedExperiment’:\n",
            "\n",
            "    trim\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:GenomicRanges’:\n",
            "\n",
            "    trim\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:IRanges’:\n",
            "\n",
            "    trim\n",
            "\n",
            "\n",
            "\n",
            "Attaching package: ‘data.table’\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:SummarizedExperiment’:\n",
            "\n",
            "    shift\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:GenomicRanges’:\n",
            "\n",
            "    shift\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:IRanges’:\n",
            "\n",
            "    shift\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:S4Vectors’:\n",
            "\n",
            "    first, second\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:lubridate’:\n",
            "\n",
            "    hour, isoweek, mday, minute, month, quarter, second, wday, week,\n",
            "    yday, year\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:purrr’:\n",
            "\n",
            "    transpose\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:scutilsR’:\n",
            "\n",
            "    %notin%\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:dplyr’:\n",
            "\n",
            "    between, first, last\n",
            "\n",
            "\n"
          ]
        }
      ],
      "source": [
        "library(Seurat)\n",
        "library(ggplot2)\n",
        "library(dplyr)\n",
        "library(scutilsR)\n",
        "library(tidyverse)\n",
        "library(celda)\n",
        "library(sceasy)\n",
        "library(enrichR)\n",
        "library(Nebulosa)\n",
        "library(magrittr)\n",
        "library(SeuratWrappers)\n",
        "library(glue)\n",
        "library(data.table)\n",
        "library(ggsci)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 21,
      "id": "3bc3354c",
      "metadata": {
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "#### 导入数据 ####\n",
        "samples <- list.dirs(\"/home/data/tanglei/project/prostate_altas/data/solo_matrix\", full.names = F, recursive = F)\n",
        "seu.list <- pbapply::pblapply(samples, function(sn) {\n",
        "  counts <- Read10X(file.path(\"/home/data/tanglei/project/prostate_altas/data/solo_matrix/\", sn, \"filtered\"))\n",
        "  sn <- gsub(\"_\", \"-\", sn) # 注意\"_\"在`CreateSeuratObject()`里有特殊的意义\n",
        "  colnames(counts) <- paste(sn, colnames(counts), sep = \"_\")\n",
        "  seu <- CreateSeuratObject(counts = counts)\n",
        "  return(seu)\n",
        "})\n",
        "## 合并样本\n",
        "seu <- base::Reduce(f = merge, x = seu.list)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "id": "3e6da20d",
      "metadata": {
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "## 粗过过滤nFeature<500\n",
        "seu <- subset(seu, nFeature_RNA > 500)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 4,
      "id": "3c708977",
      "metadata": {
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "scobj <- seu\n",
        "rm(seu)\n",
        "rm(seu.list)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 5,
      "id": "3626fb2c",
      "metadata": {
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "scobj = JoinLayers(scobj)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "b45d7b52",
      "metadata": {
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "## 计算质控指标\n",
        "##线粒体\n",
        "scobj[[\"percent.mt\"]] <- PercentageFeatureSet(scobj, pattern = \"^MT-\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "7abc327f",
      "metadata": {
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "## 细胞周期\n",
        "scobj <- NormalizeData(scobj)\n",
        "s.genes <- cc.genes.updated.2019$s.genes\n",
        "g2m.genes <- cc.genes.updated.2019$g2m.genes\n",
        "scobj <- CellCycleScoring(scobj, s.features = s.genes, g2m.features = g2m.genes)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "94be9d16",
      "metadata": {
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "## 双细胞（Seurat v5 使用 layer，不再有 @counts slot）\n",
        "library(scDblFinder)\n",
        "library(SingleCellExperiment)\n",
        "\n",
        "# 兼容 Seurat v4/v5：优先从 layer 取 counts；若存在多个 counts.* layer，先合并\n",
        "counts <- NULL\n",
        "rna <- scobj[[\"RNA\"]]\n",
        "\n",
        "layers <- tryCatch(SeuratObject::Layers(rna), error = function(e) NULL)\n",
        "if (!is.null(layers)) {\n",
        "  cnt_layers <- grep(\"^counts\", layers, value = TRUE)\n",
        "  if (length(cnt_layers) == 1) {\n",
        "    counts <- SeuratObject::GetAssayData(scobj, assay = \"RNA\", layer = cnt_layers[[1]])\n",
        "  } else if (length(cnt_layers) > 1) {\n",
        "    if (\"JoinLayers\" %in% getNamespaceExports(\"SeuratObject\")) {\n",
        "      # JoinLayers() 要求 layers 与 new 长度一致；用重复的 new 来把多个 counts.* 分组到同一个 \"counts\"\n",
        "      scobj <- SeuratObject::JoinLayers(\n",
        "        scobj,\n",
        "        assay = \"RNA\",\n",
        "        layers = cnt_layers,\n",
        "        new = rep(\"counts\", length(cnt_layers))\n",
        "      )\n",
        "      counts <- SeuratObject::GetAssayData(scobj, assay = \"RNA\", layer = \"counts\")\n",
        "    } else {\n",
        "      counts_list <- lapply(cnt_layers, function(ly) SeuratObject::GetAssayData(scobj, assay = \"RNA\", layer = ly))\n",
        "      counts <- Reduce(Matrix::cbind2, counts_list)\n",
        "    }\n",
        "  }\n",
        "}\n",
        "\n",
        "if (is.null(counts)) {\n",
        "  # Seurat v4 兜底（slot 写法）\n",
        "  counts <- SeuratObject::GetAssayData(scobj, assay = \"RNA\", slot = \"counts\")\n",
        "}\n",
        "\n",
        "sce <- SingleCellExperiment(list(counts = counts))\n",
        "sce <- scDblFinder(sce)\n",
        "\n",
        "scobj$scDblFinder.class <- sce$scDblFinder.class\n",
        "scobj$scDblFinder.score <- sce$scDblFinder.score\n",
        "\n",
        "rm(sce, counts, rna, layers)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "08ba9e5a",
      "metadata": {
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "## QC score\n",
        "QC_genes <- c(\"ACADL\", \"AP2S1\", \"ATP5F1D\", \"ATP5F1E\", \"ATP5MC1\", \"ATP5MC3\", \n",
        "              \"ATP5PF\", \"ATP5MF\", \"ATP5ME\", \"ATP5MG\", \"ATP6V1F\", \n",
        "              \"CHCHD10\", \"COA3\", \"COX5B\", \"COX6A1\", \"COX6B1\", \n",
        "              \"COX6C\", \"COX7A2\", \"COX7A2L\", \"COX7B\", \"CYCS\", \n",
        "              \"EDF1\", \"EEF1B2\", \"EIF5A\", \"FAU\", \"FKBP3\", \"FTL\", \n",
        "              \"GUK1\", \"HEPH\", \"HRAS\", \"MIF\", \"MRAP\", \"NACA\", \n",
        "              \"NDUFA1\", \"NDUFA2\", \"NDUFA4\", \"NDUFA5\", \"NDUFB7\", \n",
        "              \"NDUFC1\", \"NDUFS7\", \"NDUFV3\", \"NECAP1\", \"NLRP4\", \n",
        "              \"PDXP\", \"PFN2\", \"POLR2M\", \"RAB3A\", \"RTL8A\", \n",
        "              \"SLC16A2\", \"SNRPD2\", \"SNU13\", \"TAF1C\", \"TIMM8B\", \n",
        "              \"TPT1\", \"UBB\", \"UQCR11\", \"UQCRB\", \"UQCRQ\", \"USP50\")\n",
        "expr_matrix <- GetAssayData(scobj, assay = \"RNA\", slot = \"data\")\n",
        "log_expr_values <- log1p(expr_matrix[QC_genes, ])\n",
        "qc_scores <- colSums(log_expr_values, na.rm = TRUE)\n",
        "scobj$qc_score <- qc_scores"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 10,
      "id": "217058c7",
      "metadata": {
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "## 应激\n",
        "hot_shock <- c(\"FOS\", \"CXCL2\", \"ZFP36\", \"FOSB\", \"DUSP1\", \"ATF3\", \"CXCL8\", \n",
        "  \"NR4A1\", \"CXCL3\", \"PPP1R15A\", \"JUNB\", \"EGR1\", \"HSPA1A\", \"HSPA1B\", \n",
        "  \"SOCS3\", \"KLF6\", \"JUN\", \"IER2\", \"CXCL1\", \"NFKBIA\", \"HSPA6\", \"DNAJB1\", \n",
        "  \"IER3\", \"CCNL1\", \"MTRNR2L2\", \"IER5\", \"ID1\", \"CEBPD\", \"KRT6A\", \n",
        "  \"CYR61\", \"DEPP1\", \"CLDN4\", \"IRF1\", \"DUSP2\", \"BTG2\", \"PLAUR\", \n",
        "  \"MAFF\", \"KLF4\", \"PHLDA2\", \"TNFAIP3\", \"ACTG1\", \"BTG1\", \"DNAJB4\", \n",
        "  \"ERRFI1\", \"H3F3B\", \"HSPB1\", \"PCF11\", \"PXDC1\", \"SDC4\", \"SRF\", \n",
        "  \"TPM3\", \"USP2\", \"GADD45G\", \"ANKRD1\", \"FAM132B\", \"HIPK3\", \"HSPH1\", \n",
        "  \"IRF8\", \"KLF9\", \"NFKBIZ\", \"PDE4B\", \"RAP1B\", \"SERPINE1\", \"TPPP3\", \n",
        "  \"WAC\", \"HSPE1\", \"ARID5A\", \"DCN\", \"DUSP8\", \"HSP90AA1\", \"ID3\", \n",
        "  \"ITPKC\", \"LITAF\", \"NOP58\", \"PER1\", \"RASSF1\", \"SKIL\", \"SRSF7\", \n",
        "  \"TRA2A\", \"ZC3H12A\", \"CCRN4L\", \"DDX3X\", \"HSP90AB1\", \"IDI1\", \"LMNA\", \n",
        "  \"MYADM\", \"NPPC\", \"PHLDA1\", \"RHOB\", \"SLC10A6\", \"STAT3\", \"TRA2B\", \n",
        "  \"ZFAND5\", \"KCNE4\", \"ATF4\", \"CEBPB\", \"DDX5\", \"EGR2\", \"FOSL2\", \n",
        "  \"MYC\", \"PNP\", \"RHOH\", \"SLC38A2\", \"TAGLN2\", \"TRIB1\", \"BAG3\", \"DES\", \n",
        "  \"GADD45A\", \"JUND\", \"MAFK\", \"MYD88\", \"ODC1\", \"PNRC1\", \"RIPK1\", \n",
        "  \"SLC41A1\", \"TIPARP\", \"TUBB4B\", \"ZFP36L1\", \"BHLHE40\", \"CEBPG\", \n",
        "  \"DNAJA1\", \"EIF5\", \"GCC1\", \"HSPA5\", \"IFRD1\", \"KLF2\", \"MCL1\", \"NCKAP5L\", \n",
        "  \"OSGIN1\", \"SAT1\", \"TUBB6\", \"ZFP36L2\", \"BRD2\", \"CSRNP1\", \"ERF\", \n",
        "  \"GEM\", \"HSPA8\", \"IL6\", \"MIDN\", \"NCOA7\", \"OXNAD1\", \"SBNO2\", \"SQSTM1\", \n",
        "  \"TNFAIP6\", \"UBC\", \"ZYX\", \"MIR22HG\", \"MT1A\", \"SRSF5\", \"MT2A\", \n",
        "  \"EIF1\", \"PPP1CC\", \"ACTB\", \"ADAMTS1\", \"ADAMTS9\", \"AHNAK\", \"ANKRD11\", \n",
        "  \"ARF4\", \"AZIN1\", \"BAIAP2\", \"BAZ1A\", \"CAMK1D\", \"CCDC138\", \"CDKN1A\", \n",
        "  \"CHD4\", \"CHKA\", \"CLIC4\", \"CMSS1\", \"COL1A1\", \"CTNNB1\", \"CX3CR1\", \n",
        "  \"ELF2\", \"EP400\", \"ERN1\", \"ETF1\", \"FBXL18\", \"FLT1\", \"GADD45B\", \n",
        "  \"GLS\", \"GNAS\", \"GSK3A\", \"GSN\", \"HIVEP2\", \"INTS6\", \"JAK1\", \"JDP2\", \n",
        "  \"KDM6B\", \"KPNA1\", \"LSMEM1\", \"LUZP1\", \"MAGI3\", \"MAN1A1\", \"MAPKAPK2\", \n",
        "  \"MAPRE1\", \"MED13\", \"MSN\", \"MYLIP\", \"NABP1\", \"NASP\", \"NUFIP2\", \n",
        "  \"NUP210L\", \"PEAK1\", \"PECAM1\", \"POLG2\", \"PPP1CB\", \"PRKCG\", \"RNF19B\", \n",
        "  \"RTN4\", \"SERTAD2\", \"SGPL1\", \"SIK3\", \"SPAG9\", \"TAF4B\", \"TEX14\", \n",
        "  \"TOB2\", \"TOP1\", \"DIAPH1\", \"NEAT1\", \"PTMA\", \"ARIH1\")\n",
        "hot_shock_genes <- intersect(hot_shock, rownames(scobj))\n",
        "scobj[[\"percent.stress\"]] <- PercentageFeatureSet(scobj, features = hot_shock_genes)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 11,
      "id": "e683b910",
      "metadata": {
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "##红细胞基因\n",
        "erythrocyte_genes <- c(\n",
        "  \"GATA1\",    # 红细胞转录因子\n",
        "  \"EPOR\",     # 红细胞生成素受体\n",
        "  \"HBB\",      # 血红蛋白β链\n",
        "  \"HBA1\",     # 血红蛋白α链1\n",
        "  \"HBA2\",     # 血红蛋白α链2\n",
        "  \"KLF1\",     # 红细胞发育关键转录因子\n",
        "  \"SLC4A1\",   # 红细胞膜蛋白\n",
        "  \"EKLF\",     # 红细胞特异性转录因子\n",
        "  \"ALAS2\",    # 红细胞δ氨基酮戊酸合成酶\n",
        "  \"TAL1\"      # T细胞白血病/淋巴瘤蛋白1\n",
        ")\n",
        "erythrocyte_genes <- intersect(erythrocyte_genes, rownames(scobj))\n",
        "scobj[[\"percent.ery\"]] <- PercentageFeatureSet(scobj, features = erythrocyte_genes)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 12,
      "id": "6408bc10",
      "metadata": {
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "##计算核糖体基因\n",
        "ribo.genes <- ProjectSVR::ribo.genes\n",
        "ribo.genes <- intersect(ribo.genes, rownames(scobj))\n",
        "scobj[[\"percent.ribo\"]] <- PercentageFeatureSet(scobj, features = ribo.genes)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "1eb74428",
      "metadata": {
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "## RNA污染（Seurat v5 兼容）\n",
        "library(SingleCellExperiment)\n",
        "\n",
        "QuickCluster <- function(object) {\n",
        "  object <- NormalizeData(object)\n",
        "  object <- FindVariableFeatures(object, nfeatures = 2000)\n",
        "  object <- ScaleData(object)\n",
        "  object <- RunPCA(object)\n",
        "  object <- FindNeighbors(object, reduction = \"pca\", dims = 1:30)\n",
        "  object <- FindClusters(object)\n",
        "  return(object)\n",
        "}\n",
        "\n",
        "# 兼容 Seurat v4/v5：稳健提取 counts 矩阵\n",
        "get_counts_layer <- function(obj, assay = \"RNA\") {\n",
        "  rna <- obj[[assay]]\n",
        "  layers <- tryCatch(SeuratObject::Layers(rna), error = function(e) NULL)\n",
        "\n",
        "  if (!is.null(layers)) {\n",
        "    cnt_layers <- grep(\"^counts\", layers, value = TRUE)\n",
        "    if (length(cnt_layers) == 1) {\n",
        "      return(SeuratObject::GetAssayData(obj, assay = assay, layer = cnt_layers[[1]]))\n",
        "    }\n",
        "    if (length(cnt_layers) > 1) {\n",
        "      obj <- SeuratObject::JoinLayers(\n",
        "        obj,\n",
        "        assay = assay,\n",
        "        layers = cnt_layers,\n",
        "        new = rep(\"counts\", length(cnt_layers))\n",
        "      )\n",
        "      return(SeuratObject::GetAssayData(obj, assay = assay, layer = \"counts\"))\n",
        "    }\n",
        "  }\n",
        "\n",
        "  # Seurat v4 fallback\n",
        "  SeuratObject::GetAssayData(obj, assay = assay, slot = \"counts\")\n",
        "}\n",
        "\n",
        "seu.list <- SplitObject(scobj, split.by = \"orig.ident\")\n",
        "seu.list <- lapply(seu.list, QuickCluster)\n",
        "\n",
        "clusters <- lapply(seu.list, function(xx) xx$seurat_clusters) %>% base::Reduce(c, .)\n",
        "scobj$quick_clusters <- clusters[colnames(scobj)]\n",
        "\n",
        "# 分样本运行 decontX，避免旧版函数对 Seurat v5 slot 的兼容问题\n",
        "contam_list <- lapply(names(seu.list), function(sn) {\n",
        "  obj <- seu.list[[sn]]\n",
        "  counts <- get_counts_layer(obj, assay = \"RNA\")\n",
        "  clus <- obj$seurat_clusters\n",
        "\n",
        "  sce <- SingleCellExperiment::SingleCellExperiment(list(counts = counts))\n",
        "  sce <- celda::decontX(sce, z = as.factor(clus), verbose = FALSE)\n",
        "\n",
        "  contam <- as.numeric(SummarizedExperiment::colData(sce)$decontX_contamination)\n",
        "  names(contam) <- colnames(sce)\n",
        "  contam\n",
        "})\n",
        "\n",
        "decontX_contamination <- unlist(contam_list, use.names = TRUE)\n",
        "\n",
        "# 按 scobj 细胞顺序对齐\n",
        "contam_full <- rep(NA_real_, ncol(scobj))\n",
        "names(contam_full) <- colnames(scobj)\n",
        "contam_full[names(decontX_contamination)] <- decontX_contamination\n",
        "\n",
        "scobj$decontX_contamination <- contam_full\n",
        "\n",
        "rm(seu.list, clusters, contam_list, decontX_contamination, contam_full, get_counts_layer)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "05debc28",
      "metadata": {
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "## 计算内含子占比（从 STARsolo velocyto 输出读取）\n",
        "solo_velo_dir <- \"/home/data/tanglei/project/prostate_altas/data/solo_velocyto\"\n",
        "velo_samples <- list.dirs(solo_velo_dir, recursive = FALSE, full.names = FALSE)\n",
        "\n",
        "read_velo_mat <- function(sample_id, mat_name, base_dir) {\n",
        "  d <- file.path(base_dir, sample_id, \"filtered\")\n",
        "  m <- Matrix::readMM(file.path(d, paste0(mat_name, \".mtx\")))\n",
        "  m <- methods::as(m, \"dgCMatrix\")\n",
        "\n",
        "  features <- data.table::fread(file.path(d, \"features.tsv\"), header = FALSE)\n",
        "  barcodes <- data.table::fread(file.path(d, \"barcodes.tsv\"), header = FALSE)$V1\n",
        "\n",
        "  gene_names <- if (ncol(features) >= 2) features[[2]] else features[[1]]\n",
        "  rownames(m) <- make.unique(as.character(gene_names))\n",
        "\n",
        "  sample_tag <- gsub(\"_\", \"-\", sample_id)\n",
        "  colnames(m) <- paste0(sample_tag, \"_\", as.character(barcodes))\n",
        "  m\n",
        "}\n",
        "\n",
        "if (length(velo_samples) == 0) {\n",
        "  stop(\"No samples found under solo_velocyto directory\")\n",
        "}\n",
        "\n",
        "spliced_list <- lapply(velo_samples, function(sn) read_velo_mat(sn, \"spliced\", solo_velo_dir))\n",
        "unspliced_list <- lapply(velo_samples, function(sn) read_velo_mat(sn, \"unspliced\", solo_velo_dir))\n",
        "\n",
        "# 多样本按共同基因拼接列\n",
        "common_genes <- Reduce(intersect, lapply(spliced_list, rownames))\n",
        "common_genes <- intersect(common_genes, Reduce(intersect, lapply(unspliced_list, rownames)))\n",
        "\n",
        "spliced_all <- do.call(cbind, lapply(spliced_list, function(m) m[common_genes, , drop = FALSE]))\n",
        "unspliced_all <- do.call(cbind, lapply(unspliced_list, function(m) m[common_genes, , drop = FALSE]))\n",
        "\n",
        "# 与 Seurat 细胞名对齐\n",
        "common_cells <- intersect(colnames(scobj), colnames(spliced_all))\n",
        "common_cells <- intersect(common_cells, colnames(unspliced_all))\n",
        "missing_cells <- setdiff(colnames(scobj), common_cells)\n",
        "\n",
        "emat <- spliced_all[, common_cells, drop = FALSE]\n",
        "nmat <- unspliced_all[, common_cells, drop = FALSE]\n",
        "\n",
        "e_sum <- Matrix::colSums(emat)\n",
        "n_sum <- Matrix::colSums(nmat)\n",
        "denom <- e_sum + n_sum\n",
        "\n",
        "# 仅对匹配到 velocyto 的细胞计算；未匹配细胞默认设为 0（并给出提示）\n",
        "percent.intron <- rep(0, ncol(scobj))\n",
        "names(percent.intron) <- colnames(scobj)\n",
        "ratio_common <- ifelse(denom > 0, n_sum / denom, 0)\n",
        "percent.intron[common_cells] <- ratio_common\n",
        "\n",
        "message(\"Matched cells for intron ratio: \", length(common_cells),\n",
        "        \"; missing in velocyto: \", length(missing_cells))\n",
        "if (length(missing_cells) > 0) {\n",
        "  message(\"Example missing cells: \", paste(head(missing_cells, 5), collapse = \", \"))\n",
        "}\n",
        "\n",
        "scobj[[\"percent.intron\"]] <- percent.intron\n",
        "\n",
        "rm(solo_velo_dir, velo_samples, spliced_list, unspliced_list, common_genes,\n",
        "   spliced_all, unspliced_all, common_cells, emat, nmat, e_sum, n_sum,\n",
        "   denom, percent.intron, ratio_common, read_velo_mat)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "b8a6497b",
      "metadata": {
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "## 通过质控矩阵得到质控得分\n",
        "## 质控矩阵\n",
        "table(scobj@meta.data$orig.ident)\n",
        "sample = \"GSM4773521\"\n",
        "seu1=subset(scobj, subset = orig.ident==sample)\n",
        "\n",
        "metadata=seu1@meta.data\n",
        "qc.names = c(\"nCount_RNA\", \"nFeature_RNA\", \"percent.mt\", \"percent.ribo\", \"percent.intron\",\n",
        "             \"percent.ery\", \"scDblFinder.score\",\n",
        "             \"decontX_contamination\", \"qc_score\",\"percent.stress\")\n",
        "\n",
        "qc.mat = metadata[, qc.names, drop = FALSE]"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "f179dcf2",
      "metadata": {
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "## MinMax缩放\n",
        "qc.mat <- as.data.frame(lapply(qc.mat, as.numeric))\n",
        "rownames(qc.mat) <- rownames(metadata)\n",
        "qc.minmax <- as.data.frame(\n",
        "  apply(qc.mat, 2, function(x) {\n",
        "    (x - min(x, na.rm = TRUE)) / (max(x, na.rm = TRUE) - min(x, na.rm = TRUE))\n",
        "  })\n",
        ")\n",
        "summary(qc.minmax)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "6677259f",
      "metadata": {
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "## dist\n",
        "dist.mat = dist(qc.minmax[sample(1:nrow(qc.minmax), size = 2000), ])\n",
        "hist(dist.mat)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 34,
      "id": "9fb6a644",
      "metadata": {
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "## kNN graph\n",
        "nn = RANN::nn2(qc.minmax, k = 20)\n",
        "nn.idx <- nn$nn.idx[, -1]  # 去掉自身这个最近邻"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 35,
      "id": "56242a5d",
      "metadata": {
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "# 构建边表（from–to）\n",
        "edges <- do.call(rbind, lapply(1:nrow(nn.idx), function(i) {\n",
        "  from <- i\n",
        "  to <- nn.idx[i, ]\n",
        "  cbind(from = from, to = to)\n",
        "}))\n",
        "edges <- unique(t(apply(edges, 1, sort)))  # undirected 去重"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "cd534a10",
      "metadata": {
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "#  用 igraph 构建图\n",
        "library(igraph)\n",
        "g <- graph_from_edgelist(edges, directed = FALSE)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "35ad4fa5",
      "metadata": {
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "# Louvain 聚类\n",
        "set.seed(123)\n",
        "louvain.res <- cluster_louvain(g, resolution = 0.9)\n",
        "metadata=seu1@meta.data\n",
        "metadata$louvain <- factor(membership(louvain.res))\n",
        "\n",
        "prop.table(table(metadata$louvain))"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "22572967",
      "metadata": {
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "library(ggthemes)\n",
        "library(viridis)\n",
        "metadata_long <- metadata %>%\n",
        "  pivot_longer(\n",
        "    cols = c(qc_score, qc.names), \n",
        "    names_to = \"metric\", \n",
        "    values_to = \"value\"\n",
        "  )"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "35ac2635",
      "metadata": {
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "ggplot(metadata_long, aes(factor(louvain), value, fill = factor(louvain))) + \n",
        "  geom_violin(\n",
        "    scale = \"width\", \n",
        "    color = \"black\", \n",
        "    alpha = 0.7\n",
        "  ) +\n",
        "  geom_boxplot(\n",
        "    width = 0.2, \n",
        "    fill = \"white\", \n",
        "    color = \"black\",\n",
        "    outlier.shape = NA\n",
        "  ) +\n",
        "  facet_wrap(~metric, scales = \"free_y\", ncol = 1) +\n",
        "  theme_minimal() +\n",
        "  theme(\n",
        "    strip.text = element_text(face = \"bold\", size = 15),  \n",
        "    panel.grid.major.x = element_blank(),\n",
        "    panel.grid.minor.x = element_blank(),\n",
        "    panel.grid.minor.y = element_blank(),\n",
        "    axis.line = element_line(color = \"black\", size = 0.5),\n",
        "    axis.text = element_text(color = \"black\", size = 12),\n",
        "    axis.title = element_text(color = \"black\", face = \"bold\", size = 12),\n",
        "    plot.title = element_text(hjust = 0.5, face = \"bold\", size = 20) \n",
        "  ) +\n",
        "  labs(\n",
        "    x = \"Louvain Cluster\",\n",
        "    y = \"Value\",\n",
        "    title = \"QC Metrics Distribution Across Clusters\"\n",
        "  ) +\n",
        "  scale_y_continuous(expand = c(0.02, 0)) +\n",
        "  scale_fill_viridis(discrete = TRUE)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "5a9fe60c",
      "metadata": {
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "metadata=metadata[rownames(seu1@meta.data),]\n",
        "seu1@meta.data=metadata\n",
        "##可视化\n",
        "seu1 <- RunUMAP(seu1, reduction = \"qc\", dims = 1:9,reduction.name = \"umap_qc\")\n",
        "seu1 <- FindNeighbors(seu1, reduction = \"qc\", dims = 1:9, k.param = 20)\n",
        "seu1 <- FindClusters(seu1, resolution = 0.6,cluster.name = \"qc_cluster\")\n",
        "DimPlot(seu1, reduction = \"umap_qc\",label = T,group.by = \"louvain\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "258c92a3",
      "metadata": {
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "## label\n",
        "Idents(seu1) = \"louvain\"\n",
        "for(i in 1:nrow(seu1@meta.data)) {\n",
        "  if(seu1@meta.data$louvain[i] == 1) {\n",
        "    seu1@meta.data$final_state_GSE[i] <- \"Low\"\n",
        "  }\n",
        "  if(seu1@meta.data$louvain[i] == 2) {\n",
        "    seu1@meta.data$final_state_GSE[i] <- \"Low\"\n",
        "  }\n",
        "  if(seu1@meta.data$louvain[i] == 3) {\n",
        "    seu1@meta.data$final_state_GSE[i] <- \"High\"\n",
        "  }\n",
        "  if(seu1@meta.data$louvain[i] == 4) {\n",
        "    seu1@meta.data$final_state_GSE[i] <- \"Low\"\n",
        "  }\n",
        "  if(seu1@meta.data$louvain[i] == 5) {\n",
        "    seu1@meta.data$final_state_GSE[i] <- \"Low\"\n",
        "  }\n",
        "  if(seu1@meta.data$louvain[i] == 6) {\n",
        "    seu1@meta.data$final_state_GSE[i] <- \"High\"\n",
        "  }\n",
        "  if(seu1@meta.data$louvain[i] == 7) {\n",
        "    seu1@meta.data$final_state_GSE[i] <- \"Doubt\"\n",
        "  }\n",
        "  if(seu1@meta.data$louvain[i] == 8) {\n",
        "    seu1@meta.data$final_state_GSE[i] <- \"Low\"\n",
        "  }\n",
        "  if(seu1@meta.data$louvain[i] == 9) {\n",
        "    seu1@meta.data$final_state_GSE[i] <- \"Low\"\n",
        "  }\n",
        "    if(seu1@meta.data$louvain[i] == 10) {\n",
        "    seu1@meta.data$final_state_GSE[i] <- \"Low\"\n",
        "  }\n",
        "    if(seu1@meta.data$louvain[i] == 11) {\n",
        "    seu1@meta.data$final_state_GSE[i] <- \"Low\"\n",
        "  }\n",
        "}\n",
        "\n",
        "DimPlot(seu1, reduction = \"umap_qc\",label = T,group.by = \"final_state_GSE\")\n",
        "metadata=seu1@meta.data\n",
        "write.csv(metadata,glue(\"/home/data/tanglei/project/prostate_altas/output/{sample}_QC.csv\"))"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "5e0b8295",
      "metadata": {
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "##合并数据\n",
        "# 初始化一个空的列表以存储读取的数据框\n",
        "meta_list <- list()\n",
        "\n",
        "# 获取目录下的所有CSV文件\n",
        "csv_files <- list.files(\"/home/data/tanglei/project/prostate_altas/output\", pattern = \"*.csv\", full.names = TRUE)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "66057685",
      "metadata": {
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "# 循环读取每个CSV文件并加入到列表中\n",
        "for (file in csv_files) {\n",
        "  tmp <- fread(file, data.table = FALSE)  # 读取CSV文件\n",
        "  rownames(tmp) <- tmp$V1                  # 设置行名\n",
        "  tmp <- tmp[, -1]                         # 删除原始行名列\n",
        "  meta_list[[file]] <- tmp                 # 将数据框添加到列表\n",
        "}"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "946b4764",
      "metadata": {
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "# 将列表中的所有数据框合并为一个大的数据框，缺失的列用NA填充\n",
        "meta_list <- lapply(meta_list, function(df) {\n",
        "  df[,\"final_state_GSE\", drop = FALSE]  \n",
        "})\n",
        "\n",
        "qc <- bind_rows(meta_list)\n",
        "metadata=scobj@meta.data\n",
        "qc=qc[rownames(metadata),,drop = F]\n",
        "metadata=cbind(metadata,qc)\n",
        "scobj@meta.data=metadata"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "9676d387",
      "metadata": {
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "## 合并metadata信息\n",
        "meta = data.table::fread(\"/home/data/tanglei/project/prostate_altas/data/metadata.csv\")\n",
        "metadata = scobj@meta.data\n",
        "metadata$sample.ID = metadata$orig.ident\n",
        "metadata$GSE.ID = plyr::mapvalues(metadata$orig.ident,from = meta$sample.ID,to = meta$GSE.ID)\n",
        "metadata$celltype_article = plyr::mapvalues(metadata$orig.ident,from = meta$sample.ID,to = meta$celltype_article)\n",
        "metadata$celltype_article_H1 = plyr::mapvalues(metadata$orig.ident,from = meta$sample.ID,to = meta$celltype_article_H1)\n",
        "metadata$celltype_article_H2 = plyr::mapvalues(metadata$orig.ident,from = meta$sample.ID,to = meta$celltype_article_H2)\n",
        "metadata$celltype_article_H3 = plyr::mapvalues(metadata$orig.ident,from = meta$sample.ID,to = meta$celltype_article_H3)\n",
        "metadata$type = plyr::mapvalues(metadata$orig.ident,from = meta$sample.ID,to = meta$type)\n",
        "metadata$sample_from_met = plyr::mapvalues(metadata$orig.ident,from = meta$sample.ID,to = meta$sample_from_met)\n",
        "metadata$stage = plyr::mapvalues(metadata$orig.ident,from = meta$sample.ID,to = meta$stage)\n",
        "metadata$age = plyr::mapvalues(metadata$orig.ident,from = meta$sample.ID,to = meta$age)\n",
        "metadata$PSA = plyr::mapvalues(metadata$orig.ident,from = meta$sample.ID,to = meta$PSA)\n",
        "metadata$Gleason = plyr::mapvalues(metadata$orig.ident,from = meta$sample.ID,to = meta$Gleason)\n",
        "metadata$type_zone = plyr::mapvalues(metadata$orig.ident,from = meta$sample.ID,to = meta$type_zone)\n",
        "metadata$type_composition = plyr::mapvalues(metadata$orig.ident,from = meta$sample.ID,to = meta$type_composition)\n",
        "metadata$Prostate_volume = plyr::mapvalues(metadata$orig.ident,from = meta$sample.ID,to = meta$Prostate_volume)\n",
        "metadata$Prostate_volume_state = plyr::mapvalues(metadata$orig.ident,from = meta$sample.ID,to = meta$Prostate_volume_state)\n",
        "\n",
        "scobj@meta.data <- metadata"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "7915ae3c",
      "metadata": {
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "qs::qsave(scobj,\"/home/data/tanglei/project/prostate_altas/output/Seurat_QC.qs\")"
      ]
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "R 4.4.2",
      "language": "R",
      "name": "r442"
    },
    "language_info": {
      "codemirror_mode": "r",
      "file_extension": ".r",
      "mimetype": "text/x-r-source",
      "name": "R",
      "pygments_lexer": "r",
      "version": "4.4.2"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 5
}
